### DISCUSSION TITLE
Rendering the nodes in the tree format in reactflow

### DISCUSSION THREAD

**Discussion by saiGaneshAtmecs** [xyflow/xyflow#4165] [Q&A]
Hi Team,

I'm using  reactflow: "^11.10.4"

I have a huge tree and building a tree is working as expected and I'm using elkjs from auto layouting proexample
the issue is when i load the setNodes and setEdges with the custom nodes and edges at the time of rendering its gives this  nodes overlapping issue 

![image](https://github.com/xyflow/xyflow/assets/106595696/98f7605f-fcfc-47c0-ac3f-cf287f697d2e)

here is the [code sandbox](https://codesandbox.io/p/devbox/w3nzsr?migrateFrom=vj4n5m&embed=1&file=%2Fsrc%2FuseAutoLayout.ts&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clv2db5eg0006356ko70dlaio%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clv2db5eg0002356kn5aa7m7t%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clv2db5eg0004356kraknmt7k%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clv2db5eg0005356kvzw24d4b%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B52.06012530896131%252C47.93987469103869%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clv2db5eg0002356kn5aa7m7t%2522%253A%257B%2522id%2522%253A%2522clv2db5eg0002356kn5aa7m7t%2522%252C%2522activeTabId%2522%253A%2522clv2heu0b03fm356kzpsn2hzj%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv2db5ef0001356kul85j2fd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522src%252FFlow.tsx%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A28%252C%2522endLineNumber%2522%253A28%252C%2522startColumn%2522%253A23%252C%2522endColumn%2522%253A23%257D%255D%252C%2522id%2522%253A%2522clv2dhprv00pc356kyozpwnet%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FFlow.tsx%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A68%252C%2522endLineNumber%2522%253A68%252C%2522startColumn%2522%253A9%252C%2522endColumn%2522%253A100000008%257D%255D%252C%2522id%2522%253A%2522clv2fftxx0256356kkuoqbzdy%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FCustomNode.tsx%2522%252C%2522id%2522%253A%2522clv2firao02ck356k3536bxpt%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A6%252C%2522endLineNumber%2522%253A6%252C%2522startColumn%2522%253A2%252C%2522endColumn%2522%253A2%257D%255D%252C%2522id%2522%253A%2522clv2heu0b03fm356kzpsn2hzj%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clv2db5eg0005356kvzw24d4b%2522%253A%257B%2522id%2522%253A%2522clv2db5eg0005356kvzw24d4b%2522%252C%2522activeTabId%2522%253A%2522clv2hein703ds356kcwat8wmx%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5174%252C%2522id%2522%253A%2522clv2hein703ds356kcwat8wmx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clv2db5eg0004356kraknmt7k%2522%253A%257B%2522id%2522%253A%2522clv2db5eg0004356kraknmt7k%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv2db5eg0003356kziln2448%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clv2db6je000ddcf86iuybwhu%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clv2db5eg0003356kziln2448%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) 

and here is the image which I'm facing 
Note this error is coming 4 out of 10 times when I load the data to the component

**Comment by bcakmakoglu** [comment]
Your sandbox doesn't work, there's a couple errors in it :/

**Reply by saiGaneshAtmecs** [reply]
got it, 

I'm sharing another sandbox you can check it, this code is from reactflow proexample auto layout, 
I tried to make the same in codesandbox but its not working as it has some import issues so mentioning the below changes 
once the chart is rendered refresh the page twice and check the console it will be autolayouting will get into the the infinite loop 

changes are made in

 **initialElements.ts** 
node and edges changes 
and in
 **useAutoLayout.ts** 
   const { nodes: nextNodes, edges: nextEdges } = await **elkLayout**(
        nodes,
        edges,
        options
      );  
      

 [sandbox](https://codesandbox.io/p/devbox/dync6c?migrateFrom=sz6g9j&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clv2qdt8g0006356kkas9xm6k%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clv2qdt8g0002356k42rjqz64%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clv2qdt8g0004356kvpf6348n%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clv2qdt8g0005356kkxd95n7a%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clv2qdt8g0002356k42rjqz64%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv2qdt8f0001356kpydn7xt6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%257D%255D%252C%2522id%2522%253A%2522clv2qdt8g0002356k42rjqz64%2522%252C%2522activeTabId%2522%253A%2522clv2qdt8f0001356kpydn7xt6%2522%257D%252C%2522clv2qdt8g0005356kkxd95n7a%2522%253A%257B%2522id%2522%253A%2522clv2qdt8g0005356kkxd95n7a%2522%252C%2522activeTabId%2522%253A%2522clv2qlmku00hj356kegu87e0h%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clv2qlmku00hj356kegu87e0h%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clv2qdt8g0004356kvpf6348n%2522%253A%257B%2522id%2522%253A%2522clv2qdt8g0004356kvpf6348n%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv2qdt8g0003356kkgrn3jjt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clv2qdvqf000bdif85pnje9gz%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clv2qdt8g0003356kkgrn3jjt%2522%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by saiGaneshAtmecs** [reply]
I'm attaching a video to demonstrate this issue with the Reactflow Pro example.
[reactflow rerendering issue.zip](https://github.com/xyflow/xyflow/files/15006949/reactflow.rerendering.issue.zip)

**Reply by moklick** [reply]
Hey @saiGaneshAtmecs if you are a pro subscriber, please reach out via mail.

**Reply by saiGaneshAtmecs** [reply]
can you help me with the support email link for pro subscribers?

**Reply by bcakmakoglu** [reply]
You can contact the team at support@xyflow.com

**Reply by KairuiLiu** [reply]
> Hey @saiGaneshAtmecs if you are a pro subscriber, please reach out via mail.

Have you solved this problem yet? I have encountered the same problem

**Comment by jamalsoueidan** [comment]
reactflow able to handle many nodes..
https://reactflow.dev/examples/nodes/stress

Did you set the right width and height of the nodes, did you use the handle position from the autolayout?

**Comment by jamalsoueidan** [comment]
A question, does your nodes have multiply handles (sources) or only one input and output?
